### PR TITLE
Add Any type support for shape

### PR DESCRIPTION
**Pull request by AdilZouitine**
We can specify the arbitrary shape of a dimension using the Any type.
Example: `any_tensor = TensorType[3, 3, Any, torch.float32]`

I do not include the tests, to be honest, I am a bit lost on all the tests that you want. Could you clarify them for me @patrick-kidger ? 

So that I can do them in the next commit.

Edit: I guess a `test/test_any.py` and a function `test_any_dim` in `test/test_shape.py`

### CODE REVIEW

File: torchtyping/tensor_type.py [hunk 1]
```diff
@@ -86,7 +88,11 @@ def __class_getitem__(cls, item: Any) -> _AnnotatedType:
         layouts = []
         details = []
         for item_i in item:
-            if isinstance(item_i, (int, str, slice)) or item_i in (None, ...):
+            if (
+                isinstance(item_i, (int, str, slice))
+                or item_i in (None, ...)
+                or item_i is Any
```
Comment (patrick-kidger) [inline torchtyping/tensor_type.py:94]: These two lines can be simplified slightly to `item_i in (None, ..., Any)`

**Comment by patrick-kidger** [comment]
Haha, it's definitely true I went a bit overboard on the tests.

I think duplicating every test with a `-1` in [test_shape.py](https://github.com/patrick-kidger/torchtyping/blob/master/test/test_shape.py), to also test for `Any`, should be sufficient.

File: torchtyping/tensor_type.py (hunk 1 above)
Comment (AdilZouitine) [reply torchtyping/tensor_type.py:94]: You're right! I will correct it in the next commit :smile:

**Comment by AdilZouitine** [comment]
My pull is ready for code review :smile:

**Comment by AdilZouitine** [comment]
English is not my mother tongue, so don't hesitate to correct me :smile:

File: test/test_shape.py [hunk 2]
```diff
@@ -214,3 +215,60 @@ def _m1bm1_dim_checker(x: TensorType[-1, "b":4, -1]):
...
+    @typeguard.typechecked
+    def _anyany_dim_checker(x: TensorType[Any, Any]):
+        pass
+
+    @typeguard.typechecked
+    def _34any_dim_checker(x: TensorType[3, 4, Any]):
+        pass
+
+    @typeguard.typechecked
+    def _m14any_dim_checker(x: TensorType[Any, 4, Any]):
```
Comment (patrick-kidger) [inline test/test_shape.py:238]: This function should be called `_any4any_dim_checker`.

**Review by patrick-kidger — COMMENTED** [review]
That one small comment aside, this looks good to me -- if you can fix that I'd be happy to merge this. :)

File: test/test_shape.py (hunk 2 above)
Comment (AdilZouitine) [reply test/test_shape.py]: Fixed :smile:

**Comment by AdilZouitine** [comment]
I thank you for your reactivity, I will come back to you with new feature proposals when I will have spent more time using your library :smile:

**Comment by patrick-kidger** [comment]
LGTM. Thanks for contributing!

**Pull request merged**